File: Math/4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.py
class Solution:
    def gcd(self, a, b):
        while a :
            temp = a
            a = b%a
            b = temp
        return b
    def gcdSum(self, nums: list[int]) -> int:
        n = len(nums)
        prefixGcd = []
        mx = 0
        for num in nums:
            mx = max(mx,num)
            prefixGcd.append(gcd(num,mx))
        prefixGcd.sort()
        res =0
        for i in range(n//2):
            res += gcd(prefixGcd[i],prefixGcd[n-1-i])
        return res

File: Math/4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.java
class Solution {
    public long gcdSum(int[] nums) {
        int n = nums.length;
        int[] prefixGcd = new int[n];
        int mx = 0;
        for(int i=0;i<n;i++){
            if (mx < nums[i] ) mx = nums[i];
            prefixGcd[i] = gcd(mx,nums[i]);
        }
        Arrays.sort(prefixGcd);
        long res = 0;
        for(int i=0;i<n/2;i++){
            res +=gcd(prefixGcd[i],prefixGcd[n-1-i]);
        }
        return res;
    }
    int gcd(int a, int b){
        while(b != 0){
            int temp = b;
            b = a%b;
            a = temp;
        }
        return a;
    }
}